rust-src/concordium_base/src/curve_arithmetic/arkworks_instances.rs

```diff
@@ -0,0 +1,207 @@
+
+        let res =
+            F::from_bigint(big_int).ok_or(CurveDecodingError::NotInField(format!("{:?}", repr)))?;
+
+        Ok(res.into())
+    }
+}
+
+/// A wrapper type for `arkworks` group types.
+#[derive(PartialEq, Eq, Copy, Clone, fmt::Debug, derive_more::From)]
+pub struct ArkGroup<G>(pub(crate) G);
+
+impl<G: ark_ec::CurveGroup> ArkGroup<G> {
+    pub fn into_ark(&self) -> &G { &self.0 }
+}
+
+/// Serialization is implemented by delegating the functionality to the
+/// compressed affine representation of `ark_ec:CurveGroup`.
+impl<G: ark_ec::CurveGroup> Serial for ArkGroup<G> {
+    fn serial<B: crate::common::Buffer>(&self, out: &mut B) {
+        self.0
+            .into_affine()
+            .serialize_compressed(out)
+            .expect("Serialzation expected to succeed");
+    }
+}
+
+/// Deserialization is implemented by delegating the functionality to the
+/// compressed affine representation of `ark_ec:CurveGroup`.
+impl<G: ark_ec::CurveGroup> Deserial for ArkGroup<G> {
+    fn deserial<R: byteorder::ReadBytesExt>(source: &mut R) -> crate::common::ParseResult<Self> {
+        let res = G::Affine::deserialize_compressed(source)?;
+        Ok(ArkGroup(res.into()))
+    }
+}
+
+/// Curve configuration.
+///
+/// These parameters cannot be taken from the `arkworks` traits. Each `arkworks`
+/// curve should come with an implementation of this configuration trait.
+pub(crate) trait ArkCurveConfig<G: ark_ec::CurveGroup> {
+    const SCALAR_LENGTH: usize;
+    const GROUP_ELEMENT_LENGTH: usize;
+    const DOMAIN_STRING: &'static str;
+    type Hasher: ark_ec::hashing::HashToCurve<G>;
+}
+
+/// A blanket implementation of the `Curve` trait using the functionality of
+/// `ark_ec::CurveGroup` and curve configuration `ArkCurveConfig`. This gives an
+/// implementation of our `Curve` trait for `ArkGroup<F>` for any `F` that
+/// implements `ark_ec::CurveGroup`, provided an instance of `ArkCurveConfig`
+/// for that curve.
+impl<G: ark_ec::CurveGroup + ArkCurveConfig<G>> Curve for ArkGroup<G>
+where
+    <G as ark_ec::Group>::ScalarField: Serialize,
+{
+    type MultiExpType = GenericMultiExp<Self>;
+    type Scalar = ArkField<<G as ark_ec::Group>::ScalarField>;
+
+    const GROUP_ELEMENT_LENGTH: usize = G::GROUP_ELEMENT_LENGTH;
+    const SCALAR_LENGTH: usize = G::SCALAR_LENGTH;
+
+    fn zero_point() -> Self { ArkGroup(G::zero()) }
+
+    fn one_point() -> Self { ArkGroup(G::generator()) }
+
+    fn is_zero_point(&self) -> bool { self.0.is_zero() }
+
+    fn inverse_point(&self) -> Self { ArkGroup(-self.0) }
+
+    fn double_point(&self) -> Self { ArkGroup(self.0.double()) }
+
+    fn plus_point(&self, other: &Self) -> Self { ArkGroup(self.0 + other.0) }
+
+    fn minus_point(&self, other: &Self) -> Self { ArkGroup(self.0 - other.0) }
+
+    fn mul_by_scalar(&self, scalar: &Self::Scalar) -> Self { ArkGroup(self.0 * scalar.0) }
+
+    fn bytes_to_curve_unchecked<R: byteorder::ReadBytesExt>(b: &mut R) -> anyhow::Result<Self> {
```

Comment:
Yes, there is only one place in ElGamal
https://github.com/Concordium/concordium-base/blob/arkworks-integration/rust-src/concordium_base/src/elgamal/cipher.rs#L59

And it already returns `Result`, so using the checked version is fine, but it will be a bit slower.
Also in this case `Cipher::from_bytes_unchecked()` should be renamed into `Cipher::from_bytes()`, but it's a public method, maybe this is not desirable.

Reply:
I removed `Curve::bytes_to_curve_unchecked()` and  renamed `Cipher::from_bytes_unchecked()` to `Cipher::from_bytes()`; the method uses `deserial()` now instead of `Curve::bytes_to_curve_unchecked()`.